internal/serve/middleware/middleware.go

```diff
@@ -177,6 +179,70 @@ func (c cspItem) String() string {
 	return fmt.Sprintf("%s %s;", c.ContentType, strings.Join(c.Policy, " "))
 }
 
+// LoggingMiddleware is a middleware that logs requests to the logger.
+func LoggingMiddleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			mw := mutil.WrapWriter(rw)
+
+			reqCtx := req.Context()
+			logFields := log.F{
+				"req": middleware.GetReqID(reqCtx),
+			}
+			logCtx := log.Set(reqCtx, log.Ctx(reqCtx).WithFields(logFields))
+
+			tenant, err := tenant.GetTenantFromContext(reqCtx)
+			if err != nil {
+				log.Ctx(logCtx).Info("tenant cannot be derived from context")
+			}
+			if tenant != nil {
+				logFields["tenant_name"] = tenant.Name
+				logFields["tenant_id"] = tenant.ID
+				logCtx = log.Set(reqCtx, log.Ctx(reqCtx).WithFields(logFields))
+			}
+
+			logRequestStart(logCtx, req)
+			started := time.Now()
+			req = req.WithContext(logCtx)
```

Comment:
To me it'd make sense to update the request with the context prior to logging the request, right?

It may not make a difference for your current implementation, but it's the expected approach to log the start and end request with the updated context.

Reply:
Also, you can grab the context from req.Context() if you do that, removing the necessity of passing the context in `logRequestStart.`